### PR TITLE
feat: add constraint violation database exceptions

### PR DESCRIPTION
**Pull request by memleakd**
**Description**

This builds on #9979 by extending the same typed-exception approach to other common database constraint violations.

Database constraint failures can now be represented by more specific exception types when the driver provides enough information:

- `UniqueConstraintViolationException`
- `ForeignKeyConstraintViolationException`
- `NotNullConstraintViolationException`
- `CheckConstraintViolationException`

All of them extend the new `ConstraintViolationException`, which still extends `DatabaseException`. `UniqueConstraintViolationException` keeps the same public purpose and remains catchable as a `DatabaseException`.

This lets applications handle common integrity errors directly with `catch` blocks instead of inspecting raw driver-specific error codes or matching error messages.

When a driver reports an integrity constraint violation that CodeIgniter cannot safely classify more precisely, the parent `ConstraintViolationException` is used instead. This keeps the behavior conservative.

The same typed exception is also available through `getLastException()` when `DBDebug` is disabled.

I kept the classification close to the existing `UniqueConstraintViolationException` implementation, but I am happy to adjust the hierarchy or naming if maintainers prefer a different shape.

**Checklist:**

- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value (without duplication)
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Database/OCI8/Connection.php [hunk 1]
```diff
@@ -123,6 +123,33 @@ protected function isUniqueConstraintViolation(int|string $code, string $message
...
+        return in_array((int) $code, [2291, 2292], true);
+    }
+
+    /**
+     * Checks whether the native database error represents a NOT NULL constraint violation.
+     */
+    protected function isNotNullConstraintViolation(int|string $code, string $message): bool
+    {
+        // ORA-01400: cannot insert NULL.
+        return (int) $code === 1400;
```
Comment (michalsn) [inline system/Database/OCI8/Connection.php]: We should also cover the `UPDATE` case: https://docs.oracle.com/en/error-help/db/ora-01407/?r=26ai

File: system/Database/MySQLi/Connection.php [hunk 2]
```diff
@@ -106,6 +106,33 @@ protected function isUniqueConstraintViolation(int|string $code, string $message
         return $code === 1062;
     }
 
+    /**
+     * Checks whether the native database error represents a foreign key constraint violation.
+     */
+    protected function isForeignKeyConstraintViolation(int|string $code, string $message): bool
+    {
+        // ER_ROW_IS_REFERENCED_2, ER_NO_REFERENCED_ROW_2.
+        return in_array($code, [1451, 1452], true);
```
Comment (michalsn) [inline system/Database/MySQLi/Connection.php]: What about 1216/1217 codes? They seem to be documented as well: https://dev.mysql.com/doc/mysql-errors/8.0/en/server-error-reference.html

File: system/Database/MySQLi/Connection.php [hunk 3]
```diff
@@ -106,6 +106,33 @@ protected function isUniqueConstraintViolation(int|string $code, string $message
...
+        return $code === 1048;
+    }
+
+    /**
+     * Checks whether the native database error represents a CHECK constraint violation.
+     */
+    protected function isCheckConstraintViolation(int|string $code, string $message): bool
+    {
+        // ER_CHECK_CONSTRAINT_VIOLATED: check constraint is violated.
+        return $code === 3819;
```
Comment (michalsn) [inline system/Database/MySQLi/Connection.php]: Can we also support MariaDB codes? https://mariadb.com/docs/server/reference/error-codes/mariadb-error-codes-4000-to-4099/e4025

**Review by michalsn — COMMENTED** [review]
Overall, I like this direction. Could we add a few more live tests where the existing table structure allows it, particularly for `NOT NULL` updates and foreign-key violations?

File: system/Database/OCI8/Connection.php (hunk 1 above)
Comment (memleakd) [reply system/Database/OCI8/Connection.php]: Added `ORA-01407` as well, and covered it in the tests. Thanks for spotting the update case.

File: system/Database/MySQLi/Connection.php (hunk 2 above)
Comment (memleakd) [reply system/Database/MySQLi/Connection.php]: Yep, added both legacy FK codes with coverage. Thanks!

File: system/Database/MySQLi/Connection.php (hunk 3 above)
Comment (memleakd) [reply system/Database/MySQLi/Connection.php]: Added it, but with a small guard since MySQL also uses `4025` for a different error. There’s a negative test for that case too.

**Comment by memleakd** [comment]
Thanks for the review @michalsn. I pushed updates for the cases you pointed out.

While checking the surrounding driver codes, <mark>**I also added**</mark> MySQL `3673` for the same `NOT NULL` bucket as `1048`.

**One small detail**: I guarded MariaDB `4025` by message instead of treating the code alone as enough, because MariaDB documents it as `ER_CONSTRAINT_FAILED`, but MySQL uses the same numeric code for `ER_INNODB_AUTOEXTEND_SIZE_OUT_OF_RANGE`. I added a negative test for that MySQL case too.

References I checked:
- Oracle `ORA-01407`: https://docs.oracle.com/en/error-help/db/ora-01407
- MySQL `1216`, `1217`, `3673`, `4025`: https://dev.mysql.com/doc/mysql-errors/8.4/en/server-error-reference.html
- MariaDB `4025`: https://mariadb.com/docs/server/reference/error-codes/mariadb-error-codes-4000-to-4099/e4025

**Review by patel-vansh — APPROVED** [review]

File: system/Database/MySQLi/Connection.php [hunk 4]
```diff
@@ -106,6 +106,43 @@ protected function isUniqueConstraintViolation(int|string $code, string $message
...
+        }
+
+        // MariaDB reports CHECK failures as ER_CONSTRAINT_FAILED, while MySQL uses 4025 for other errors.
+        if ($code !== 4025) {
+            return false;
+        }
+
+        $message = strtolower($message);
+
+        return str_contains($message, 'constraint') && str_contains($message, 'failed');
```
Comment (michalsn) [inline system/Database/MySQLi/Connection.php]: Could we avoid matching the error message here? Some additional check like `stripos($this->getVersion(), 'MariaDB') !== false` seems more reliable.

File: system/Database/SQLSRV/Connection.php [hunk 5]
```diff
@@ -121,20 +117,65 @@ protected function isUniqueConstraintViolation(int|string $code, string $message
...
 
+    /**
+     * Checks whether the native database error represents a foreign key constraint violation.
+     */
+    protected function isForeignKeyConstraintViolation(int|string $code, string $message): bool
+    {
+        return $this->getVendorErrorCode($code) === 547
+            && $this->hasSQLState($code, '23000')
+            && str_contains($message, 'FOREIGN KEY constraint');
+    }
```
Comment (michalsn) [inline system/Database/SQLSRV/Connection.php]: I'm not sure about this change. In SQLSRV, messages can be localized, so this will not be 100% reliable - more like "best effort".

**Review by michalsn — COMMENTED** [review]
Thanks for the update.

File: system/Database/MySQLi/Connection.php (hunk 4 above)
Comment (memleakd) [reply system/Database/MySQLi/Connection.php]: Good point, thanks! Changed this to use the server version check.

File: system/Database/SQLSRV/Connection.php (hunk 5 above)
Comment (memleakd) [reply system/Database/SQLSRV/Connection.php]: Agreed, thanks! Removed this and kept `547` generic.

File: tests/system/Database/Live/ConstraintViolationExceptionTest.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        $expectedException = $this->db->DBDriver === 'SQLSRV'
+            ? ConstraintViolationException::class
+            : ForeignKeyConstraintViolationException::class;
+
+        $this->expectException($expectedException);
+
+        $this->db->table('cv_child')->insert([
+            'id'        => 1,
+            'parent_id' => 999,
+        ]);
```
Comment (michalsn) [inline tests/system/Database/Live/ConstraintViolationExceptionTest.php]: For SQLSRV we expect the generic parent, but `expectException()` also accepts its subclasses. Could we catch `DatabaseException` and use `assertSame($expectedException, $e::class)` to verify the exact classification?

File: tests/system/Database/ConstraintViolationExceptionTest.php [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+    #[DataProvider('provideCreatesConstraintViolationExceptions')]
+    public function testCreatesConstraintViolationExceptions(
+        BaseConnection $db,
+        int|string $code,
+        string $message,
+        string $expectedException,
+    ): void {
+        $exception = $db->createDatabaseException($message, $code);
+
+        $this->assertSame($exception::class, $expectedException);
```
Comment (michalsn) [inline tests/system/Database/ConstraintViolationExceptionTest.php]: Please reverse parameters.

Comment (memleakd) [reply tests/system/Database/ConstraintViolationExceptionTest.php]: Fixed, thanks!

File: tests/system/Database/Live/ConstraintViolationExceptionTest.php (hunk 6 above)
Comment (memleakd) [reply tests/system/Database/Live/ConstraintViolationExceptionTest.php]: Good catch, thanks. Updated this to assert the exact exception class.

**Review by michalsn — APPROVED** [review]
Looks solid to me. Thanks!

**Review by paulbalandan — APPROVED** [review]

**Comment by michalsn** [comment]
Thank you @memleakd

**Comment by memleakd** [comment]
Thank you @michalsn, @patel-vansh, @paulbalandan

**Review by neznaika0 — COMMENTED** [review]
@memleakd 
I suggest updating the test (provider). If all the databases are not installed in the system, I think they should be skipped as in other tests. Otherwise, we will get errors, for example
`Error: Call to undefined function CodeIgniter\Database\SQLSRV\sqlsrv_errors()`
